### PR TITLE
R4R: Add snapshot function

### PR DESCRIPTION
**Pull request by zhiqiang-bianjie**
Resolves: #2032

### CODE REVIEW

File: server/snapshot.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+const flagTmpDir = "tmp-dir"
+const pathSeparator = string(os.PathSeparator)
+
+// SnapshotCmd delete historical block data and index data
+func SnapshotCmd(ctx *Context) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "snapshot",
+		Short: "snapshot current block information and drop other block",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			home := viper.GetString("home")
```
Comment (chengwenxi) [inline server/snapshot.go]: `HomeFlag` has been defined.

File: server/snapshot.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+			fmt.Println(fmt.Sprintf("snapshot file is stored in [%s]", targetDir))
+			return nil
+		},
+	}
+	cmd.Flags().String(flagTmpDir, "", "snapshot file storage directory")
+	return cmd
+}
+
+func loadDb(name, path string) *dbm.GoLevelDB {
+	db, err := dbm.NewGoLevelDB(name, path)
```
Comment (chengwenxi) [inline server/snapshot.go]: The `GoLevelDB` is an option, there may be other dbs.

File: server/snapshot.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	}
+
+	gr, found, err := sourceWAL.SearchForEndHeight(height, &consensus.WALSearchOptions{IgnoreDataCorruptionErrors: true})
+	if err != nil {
+		return err
+	}
+	if !found {
+		return fmt.Errorf("cannot replay height %d. WAL does not contain #ENDHEIGHT for %d", height, height-1)
+	}
+	defer gr.Close()
```
Comment (chengwenxi) [inline server/snapshot.go]: move gr.Close() to `gr` is defined.

File: server/snapshot.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+)
+
+const flagTmpDir = "tmp-dir"
+const pathSeparator = string(os.PathSeparator)
+
+// SnapshotCmd delete historical block data and index data
+func SnapshotCmd(ctx *Context) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "snapshot",
+		Short: "snapshot current block information and drop other block",
```
Comment (zhangyelong) [inline server/snapshot.go]:
```suggestion
		Short: "snapshot the latest block information and drop the others",
```

**Pull request merged**